Improvement of anonymity computation

onvej-sl opened the pull request:
This PR fixes https://github.com/zkSNACKs/WalletWasabi/issues/7635 and is replacement of https://github.com/zkSNACKs/WalletWasabi/pull/7756.

@nopara73 @andrewkozlik

soosr commented:
> @soosr can you test if this PR also results in "forever loading"?

CI fails due to an exception. Can't test.

```
2022-04-28 09:14:57.469 [23] INFO       WalletManager.StartWalletAsync (164)    Starting wallet 'Normal'...
2022-04-28 09:14:57.634 [7] INFO        Wallet.Dispose (373)    Initial Transaction Processing finished in 154 milliseconds.
2022-04-28 09:14:57.636 [7] INFO        Wallet.Dispose (180)    StartAsync finished in 161 milliseconds.
2022-04-28 09:14:57.738 [7] ERROR       WalletManagerViewModel.LoadWalletAsync (157)    System.NotImplementedException: The method or operation is not implemented.
   at WalletWasabi.Blockchain.Transactions.SmartTransaction.ExtractKeyId(Script scriptPubKey) in WalletWasabi\Blockchain\Transactions\SmartTransaction.cs:line 152
   at WalletWasabi.Blockchain.Transactions.SmartTransaction.<>c.<GetForeignVirtualOutputs>b__63_1(IndexedTxOut o) in WalletWasabi\Blockchain\Transactions\SmartTransaction.cs:line 176
   at System.Linq.Lookup`2.Create(IEnumerable`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.GroupedEnumerable`2.GetEnumerator()
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.HashSet`1.UnionWith(IEnumerable`1 other)
   at System.Collections.Generic.HashSet`1..ctor(IEnumerable`1 collection, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToHashSet[TSource](IEnumerable`1 source, IEqualityComparer`1 comparer)
   at WalletWasabi.Blockchain.Transactions.SmartTransaction.GetForeignVirtualOutputs() in WalletWasabi\Blockchain\Transactions\SmartTransaction.cs:line 176
   at WalletWasabi.Blockchain.Transactions.SmartTransaction.<.ctor>b__0_4() in WalletWasabi\Blockchain\Transactions\SmartTransaction.cs:line 46
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at WalletWasabi.Blockchain.Transactions.SmartTransaction.get_ForeignVirtualOutputs() in WalletWasabi\Blockchain\Transactions\SmartTransaction.cs:line 127
   at WalletWasabi.Blockchain.Analysis.BlockchainAnalyzer.ComputeAnonymityContribution(SmartCoin transactionOutput, HashSet`1 relevantOutpoints) in WalletWasabi\Blockchain\Analysis\BlockchainAnalyzer.cs:line 123
   at WalletWasabi.Blockchain.Analysis.BlockchainAnalyzer.<>c__DisplayClass7_0.<ComputeInputSanction>g__ComputeInputSanctionHelper|1(SmartCoin transactionOutput) in WalletWasabi\Blockchain\Analysis\BlockchainAnalyzer.cs:line 112
   at WalletWasabi.Blockchain.Analysis.BlockchainAnalyzer.ComputeInputSanction(SmartTransaction analyzedTransaction, SmartCoin transactionInput) in WalletWasabi\Blockchain\Analysis\BlockchainAnalyzer.cs:line 116
   at WalletWasabi.Blockchain.Analysis.BlockchainAnalyzer.<>c__DisplayClass5_1.<AnalyzeWalletInputs>b__1(SmartCoin i) in WalletWasabi\Blockchain\Analysis\BlockchainAnalyzer.cs:line 75
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.Min(IEnumerable`1 source)
   at WalletWasabi.Blockchain.Analysis.BlockchainAnalyzer.AnalyzeWalletInputs(SmartTransaction tx, Int32& newInputAnonset, Int32& sanctionedInputAnonset) in WalletWasabi\Blockchain\Analysis\BlockchainAnalyzer.cs:line 75
   at WalletWasabi.Blockchain.Analysis.BlockchainAnalyzer.Analyze(SmartTransaction tx) in WalletWasabi\Blockchain\Analysis\BlockchainAnalyzer.cs:line 39
   at WalletWasabi.Blockchain.TransactionProcessing.TransactionProcessor.ProcessNoLock(SmartTransaction tx) in WalletWasabi\Blockchain\TransactionProcessing\TransactionProcessor.cs:line 282
   at WalletWasabi.Blockchain.TransactionProcessing.TransactionProcessor.Process(IEnumerable`1 txs) in WalletWasabi\Blockchain\TransactionProcessing\TransactionProcessor.cs:line 59
   at WalletWasabi.Wallets.Wallet.LoadWalletStateAsync(CancellationToken cancel) in WalletWasabi\Wallets\Wallet.cs:line 375
   at WalletWasabi.Wallets.Wallet.StartAsync(CancellationToken cancel) in WalletWasabi\Wallets\Wallet.cs:line 186
   at WalletWasabi.Wallets.WalletManager.StartWalletAsync(Wallet wallet) in WalletWasabi\Wallets\WalletManager.cs:line 165
   at WalletWasabi.Wallets.WalletManager.StartWalletAsync(Wallet wallet) in WalletWasabi\Wallets\WalletManager.cs:line 174
   at WalletWasabi.Fluent.ViewModels.WalletManagerViewModel.<>c__DisplayClass12_0.<<LoadWalletAsync>b__0>d.MoveNext() in WalletWasabi.Fluent\ViewModels\WalletManagerViewModel.cs:line 149
--- End of stack trace from previous location ---
   at WalletWasabi.Fluent.ViewModels.WalletManagerViewModel.LoadWalletAsync(Wallet wallet) in WalletWasabi.Fluent\ViewModels\WalletManagerViewModel.cs:line 149

```

andrewkozlik commented on WalletWasabi/Blockchain/WalletVirtualOutput.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace WalletWasabi.Blockchain;
+
+public record WalletVirtualInput
```
The file name doesn't match the class name.

onvej-sl replied:
Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/7853/commits/91d4e679598d1a9219c9c1ea42ec163f5780cbff.

onvej-sl commented:
I have one unresolved issue: I expect a single `SmartTransaction` can be analyzed multiple times but I don't expect inputs and outputs of the transaction to be added, deleted or modified between the analyses. (That is why I implemented the properties in the commits https://github.com/zkSNACKs/WalletWasabi/pull/7853/commits/15712a04c99622534aa4f1150254202a0623f742 and https://github.com/zkSNACKs/WalletWasabi/pull/7853/commits/ddaa07db7085e8e0915b44b99235d24f1c8984a9 as lazy.) Nevertheless, the test `OtherWalletChangesThings` suggest this assumption is not true. Unless the test is outdated, I'm asking: Under what circumstances can it happen that a transaction is changed on the fly? Can this happen when the transaction is being analyzed?

onvej-sl commented:
> CI fails. I can't test it due to an exception.

https://github.com/zkSNACKs/WalletWasabi/pull/7853/commits/4c0f98698e8737dd27ba73c5c22ea0d63f0b822f fixes it.

 By the way, the test `SimpleDirectClusteringAsync` uses an obsolete P2PK script type. I'm not sure whether it is an intention or a mistake.

soosr commented:
> @soosr go for another round of test please

Same. "Forever loading" happens as described [here](https://github.com/zkSNACKs/WalletWasabi/pull/7841#issuecomment-1111829634).

Additional information for both PRs.
During loading, if I close Wasabi, the process doesn't finish. I have to kill it.

onvej-sl commented:
@nopara73 I still don't understand it.
> I don't see how this could be an issue for you.

The properties of `SmartTransaction` implemented in https://github.com/zkSNACKs/WalletWasabi/commit/15712a04c99622534aa4f1150254202a0623f742 and https://github.com/zkSNACKs/WalletWasabi/commit/ddaa07db7085e8e0915b44b99235d24f1c8984a9 are computed only once. If the inputs or outputs of the transaction are changed, the properties don't return the correct value.

> OtherWalletChangesThings test ensures when another wallet is loaded that shares a transaction with the same wallet then transaction properties are changing.

Which properties can change? To be specific, can it be inputs and outputs that will be changed?

onvej-sl commented:
> Which properties can change? To be specific, can it be inputs and outputs that will be changed?

I discussed it with Andrew an he probably figured it out: If a new wallet is loaded, new inputs can be added to `WalletInputs`.

nopara73 commented:
I can verify the performance issue is fixed.

nopara73 commented:
@onvej-sl Yes, `WalletInputs` and `WalletOutputs` can be changed. `WalletOutputs` test when the second wallet is loaded adds its own inputs and outputs to these properties. And I see now this is your problem. Can you do something with it or you need my help to figure something out?

soosr commented:
> @soosr
> 
> > Same. "Forever loading" happens as described [#7841 (comment)](https://github.com/zkSNACKs/WalletWasabi/pull/7841#issuecomment-1111829634).
> 
> I expect [3df81b3](https://github.com/zkSNACKs/WalletWasabi/commit/3df81b36b64c3e4e8629523f4fb363eb8bdf9bbb) to sort it out.

I can confirm, fixed.

onvej-sl commented:
The preceding three commits fix code style.

andrewkozlik commented on WalletWasabi/Blockchain/Analysis/BlockchainAnalyzer.cs:
```diff
@@ -20,58 +21,61 @@ public BlockchainAnalyzer(int privacyLevelThreshold)
 
-		newInputAnonset = Intersect(distinctWalletInputPubKeys.Select(x => x.AnonymitySet), coefficient);
+		newInputAnonset = Intersect(tx.WalletVirtualInputs.Select(x => x.HdPubKey.AnonymitySet), coefficient);
 
```
This deserves a comment. I suggest:
If we remix with siblings from earlier CoinJoins, then we must avoid double-counting their contribution to our anonymity set. Search the history of each input to find any siblings with which we are remixing in the current transaction. Discount the anonymity gain that we got from them earlier and are now losing due to the remix.

andrewkozlik commented on WalletWasabi/Blockchain/Analysis/BlockchainAnalyzer.cs:
```diff
@@ -20,58 +21,61 @@ public BlockchainAnalyzer(int privacyLevelThreshold)
 		{
-			key.SetAnonymitySet(newInputAnonset);
+			anonsets.Add(virtualInput.SmartCoins.Select(i => Math.Max(1, virtualInput.HdPubKey.AnonymitySet - (int)coinjoinAnalyzer.ComputeInputSanction(i))).Min());
 		}
```
Here I'd also add a comment: Recompute `newInputAnonset` using the discounted anonsets. This result is used as the base for computing the anonymity of this transaction's outputs.

andrewkozlik commented on WalletWasabi/Blockchain/Analysis/BlockchainAnalyzer.cs:
```diff
@@ -99,36 +103,19 @@ private int Intersect(IEnumerable<int> anonsets, double coefficient)
 			// Anonset gain cannot be larger than others' input count.
-			var anonset = Math.Min(equalOutputCount - ownEqualOutputCount, inputCount - ownInputCount);
-
 			// Picking randomly an output would make our anonset: total/ours.
```
Let's remove the comment about picking, since that's done inside `ComputeAnonymityContribution`.

andrewkozlik commented on WalletWasabi/Blockchain/Transactions/SmartTransaction.cs:
```diff
@@ -106,8 +139,143 @@ public SmartTransaction(Transaction transaction, Height height, uint256? blockHa
+	private HashSet<WalletVirtualOutput> _walletVirtualOutputs;
+	private HashSet<ForeignVirtualOutput> _foreignVirtualOutputs;
+
+	private bool recomputeForeignInputs = true;
+	private bool recomputeForeignOutputs = true;
+
+	private bool recomputeWalletVirtualInputs = true;
+	private bool recomputeWalletVirtualOutputs = true;
+	private bool recomputeForeignVirtualOutputs = true;
```
Rather than having these recompute flags, wouldn't it be possible to assign `null` to the corresponding variable when it needs to be recomputed? Or C# doesn't allow that?

onvej-sl replied:
Definitely possible and better than my approach.

Done in https://github.com/zkSNACKs/WalletWasabi/pull/7853/commits/35ae3a5840462806726fe9a4c1e7b7897c98ec33 and https://github.com/zkSNACKs/WalletWasabi/pull/7853/commits/3333318d24c80a0f2fc18f87e806594146541a0d.

andrewkozlik commented on WalletWasabi/Blockchain/Analysis/CoinjoinAnalyzer.cs:
```diff
@@ -0,0 +1,55 @@
+
+		return ComputeInputSanctionHelper(transactionInput);
+	}
+
```
Let's add some documentation: Computes how much the foreign outputs of AnalyzedTransaction contribute to the anonymity of our transactionOutput. Sometimes we are only interested in how much a certain subset of foreign outputs contributed. This subset can be specified in relevantOutpoints, otherwise all outputs are considered relevant.

andrewkozlik commented on WalletWasabi/Blockchain/Analysis/CoinjoinAnalyzer.cs:
```diff
@@ -0,0 +1,55 @@
+
+		Money amount = walletVirtualOutputs.Where(o => o.Outpoints.Contains(transactionOutput.OutPoint)).First().Amount;
+		Func<ForeignVirtualOutput, bool> isRelevantVirtualOutput = output => relevantOutpoints is null ? true : relevantOutpoints.Intersect(output.Outpoints).Any();
+
```
Here I'd suggest a comment "Count the outputs that have the same value as our transactionOutput."

andrewkozlik commented on WalletWasabi/Blockchain/Analysis/CoinjoinAnalyzer.cs:
```diff
@@ -0,0 +1,55 @@
+
+		var equalValueWalletVirtualOutputCount = walletVirtualOutputs.Where(o => o.Amount == amount).Count();
+		var equalValueForeignRelevantVirtualOutputCount = foreignVirtualOutputs.Where(o => o.Amount == amount).Where(isRelevantVirtualOutput).Count();
+
```
And here I'd provide the following explanation:
The anonymity set should increase by the number of equal-valued foreign ouputs. If we have multiple equal-valued wallet outputs, then we divide the increase evenly between them. The rationale behind this is that picking randomly an output would make our anonset: total/ours = 1 + foreign/ours, so the increase in anonymity is foreign/ours.

nopara73 commented:
Closing this in favor of https://github.com/zkSNACKs/WalletWasabi/pull/7932
